src/main/scala/com/microsoft/ml/spark/explainers/ImageSHAP.scala

```diff
@@ -0,0 +1,121 @@
+import org.apache.spark.ml.ComplexParamsReadable
+import org.apache.spark.ml.image.ImageSchema
+import org.apache.spark.ml.linalg.SQLDataTypes.VectorType
+import org.apache.spark.ml.linalg.{Vector => SV}
+import org.apache.spark.ml.param.shared.HasInputCol
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.sql.functions.{col, explode}
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row}
+
+import java.awt.image.BufferedImage
+
+trait ImageSHAPParams
+  extends KernelSHAPParams
+    with HasCellSize
+    with HasModifier
+    with HasInputCol
+    with HasSuperpixelCol {
+  self: ImageSHAP =>
+
+  def setInputCol(value: String): this.type = this.set(inputCol, value)
+
+  setDefault(cellSize -> 16, modifier -> 130, superpixelCol -> "superpixels")
+}
+
+class ImageSHAP(override val uid: String)
+  extends KernelSHAPBase(uid)
+    with ImageSHAPParams
+    with ImageExplainer {
+
+  logClass()
+
+  def this() = {
+    this(Identifiable.randomUID("ImageSHAP"))
+  }
+
+  private def sample(bi: BufferedImage, spd: SuperpixelData, numSamplesOpt: Option[Int]): Seq[(ImageFormat, SV)] = {
+    val effectiveNumSamples = KernelSHAPBase.getEffectiveNumSamples(numSamplesOpt, spd.clusters.size)
+    val sampler = new KernelSHAPImageSampler(bi, spd, effectiveNumSamples)
+    (1 to effectiveNumSamples).map {
+      _ =>
+        val (outputImage, feature, _) = sampler.sample
+        val (path, height, width, nChannels, mode, decoded) = ImageUtils.toSparkImageTuple(outputImage)
+        val imageFormat = ImageFormat(path, height, width, nChannels, mode, decoded)
+        (imageFormat, feature)
+    }
+  }
+
+  private lazy val numSampleOpt = this.getNumSamplesOpt
+
+  private lazy val imageSamplesUdf = {
+    UDFUtils.oldUdf(
+      {
+        (image: Row, sp: Row) =>
+          val bi = ImageUtils.toBufferedImage(image)
+          val spd = SuperpixelData.fromRow(sp)
+          sample(bi, spd, numSampleOpt)
+      },
+      getSampleSchema(ImageSchema.columnSchema)
+    )
+  }
+
+  private lazy val binarySamplesUdf = {
+    UDFUtils.oldUdf(
+      {
+        (data: Array[Byte], sp: Row) =>
+          val biOpt = ImageUtils.safeRead(data)
+          val spd = SuperpixelData.fromRow(sp)
+          biOpt.map {
+            bi =>
+              sample(bi, spd, numSampleOpt)
+          }.getOrElse(Seq.empty)
+      },
+      getSampleSchema(ImageSchema.columnSchema)
+    )
+  }
+
+  override protected def createSamples(df: DataFrame, idCol: String, coalitionCol: String): DataFrame = {
+    val samplingUdf = df.schema(getInputCol).dataType match {
+      case BinaryType =>
+        this.binarySamplesUdf
+      case t if ImageSchemaUtils.isImage(t) =>
+        this.imageSamplesUdf
+    }
+
+    df.withColumn("samples", explode(samplingUdf(col(getInputCol), col(getSuperpixelCol))))
+      .select(
+        col(idCol),
+        col("samples.coalition").alias(coalitionCol),
+        col("samples.sample").alias(getInputCol)
```

Comment:
I see some of these strings in a few places, might be a good idea to put them in some constant file and reference from there

Reply:
Yes agree with this, someplace that crystalizes this new API in a SparkType might reduce duplication and clarify things

Reply:
done.